README.md
Resque Meta
===========

[![Build Status](https://secure.travis-ci.org/lmarlow/resque-meta.png)](http://travis-ci.org/lmarlow/resque-meta)

A [Resque][rq] plugin. Requires Resque 1.8 and Redis 2+.

If you want to be able to add metadata for a job
to track anything you want, extend it with this module.

For example:

    require 'resque-meta'

    class MyJob
      extend Resque::Plugins::Meta

      def self.perform(meta_id, *args)
        heavy_lifting
      end
    end

    meta0 = MyJob.enqueue('stuff')
    meta0.enqueued_at # => 'Wed May 19 13:42:41 -0600 2010'
    meta0.meta_id # => '03c9e1a045ad012dd20500264a19273c'
    meta0['foo'] = 'bar' # => 'bar'
    meta0.save

    # later
    meta1 = MyJob.get_meta('03c9e1a045ad012dd20500264a19273c')
    meta1.job_class # => MyJob
    meta1.enqueued_at # => 'Wed May 19 13:42:41 -0600 2010'
    meta1['foo'] # => 'bar'

[rq]: http://github.com/resque/resque


HISTORY.md
## 2.0.1 (2013-11-25)

* Include Resque::Helpers for [breaking change in resque 1.25][resque125] - [jsanders][jsanders] [Pull Request][pull13], [mcfiredrill][mcfiredrill] [Pull Request][pull12]

## 2.0.0 (2012-12-04)

* Use redis setex to expire metadata from Pascal Brogle - [broglep-koubachi][broglep-koubachi] [Pull Request][pull9]
* setex require redis 2+, hence the major version bump

## 1.0.3 (2011-02-10)

* Document and test Metadata#save from Kurt Werle - [kwerle][kwerle] [Pull Request][pull3]
* Metadata#seconds_enqueued and #seconds_processing are more precise from [Steve Howell][showell]

## 1.0.2 (2010-11-03)

* Use a more random default meta_id with SHA1 from Cory Forsyth - [bantic][bantic] [Pull Request][pull2]

## 1.0.1 (2010-09-10)

* Add lib/resque-meta.rb for convenient requiring
* Doc fixes from Jeffrey Chupp - [semanticart][semanticart] [Pull Request][pull1]
* Depend more strictly on resque (>= 1.8, < 2.0)

## 1.0.0 (2010-06-04)

* Initial release

[bantic]: https://github.com/bantic
[showell]: http://librelist.com/browser//resque/2010/12/8/recording-time-in-queue-and-time-to-process/#1de6433232ac1264286feeed1f8f219e
[semanticart]: https://github.com/semanticart
[kwerle]: https://github.com/kwerle
[broglep-koubachi]: https://github.com/broglep-koubachi
[pull1]: https://github.com/lmarlow/resque-meta/pull/1
[pull2]: https://github.com/lmarlow/resque-meta/pull/2
[pull3]: https://github.com/lmarlow/resque-meta/pull/3
[pull9]: https://github.com/lmarlow/resque-meta/pull/9
[resque125]: https://github.com/resque/resque/issues/1150
[jsanders]: https://github.com/jsanders
[mcfiredrill]: https://github.com/mcfiredrill
[pull12]: https://github.com/lmarlow/resque-meta/pull/12
[pull13]: https://github.com/lmarlow/resque-meta/pull/13


test/meta_test.rb
require File.dirname(__FILE__) + '/test_helper'
require 'resque-meta'

class MetaJob
  extend Resque::Plugins::Meta
  @queue = :test

  def self.expire_meta_in
    1
  end

  def self.perform(meta_id, key, val)
    meta = get_meta(meta_id)
    meta[key] = val
    meta.save
  end
end

class AnotherJob
  extend Resque::Plugins::Meta
  @queue = :test

  def self.perform(meta_id)
  end
end

class SlowJob
  extend Resque::Plugins::Meta
  @queue = :test

  def self.expire_meta_in
    1
  end

  def self.perform(meta_id, key, val)
    meta = get_meta(meta_id)
    meta[key] = val
    meta.save
    sleep 1
  end
end

class FailingJob
  extend Resque::Plugins::Meta
  @queue = :test

  def self.perform(*args)
    raise 'boom'
  end
end

class TransientJob
  extend Resque::Plugins::Meta
  @queue = :test

  def self.perform(meta_id)
  end

  def self.before_finish_expire_in
    1
  end
end

class MetaTest < Test::Unit::TestCase
  def setup
    Resque.redis.flushall
  end

  def test_meta_version
    assert_equal '2.0.1', Resque::Plugins::Meta::Version
  end

  def test_lint
    assert_nothing_raised do
      Resque::Plugin.lint(Resque::Plugins::Meta)
    end
  end

  def test_resque_version
    major, minor, patch = Resque::Version.split('.')
    assert_equal 1, major.to_i
    assert minor.to_i >= 8
  end

  def test_enqueued_metadata
    now = Time.now
    meta = MetaJob.enqueue('foo', 'bar')
    assert_not_nil meta
    assert_not_nil meta.meta_id
    assert meta.enqueued_at.to_f > now.to_f, "#{meta.enqueued_at} should be after #{now}"
    assert meta.seconds_enqueued > 0.0, "seconds_enqueued should be greater than zero"
    assert meta.enqueued?
    assert !meta.started?
    assert_equal 0, meta.seconds_processing
    assert !meta.finished?
    assert_nil meta['foo']
    assert_equal Resque::Plugins::Meta::Metadata, meta.class
    assert_equal MetaJob, meta.job_class
  end

  def test_processed_job
    meta = MetaJob.enqueue('foo', 'bar')
    assert_nil meta['foo']
    worker = Resque::Worker.new(:test)
    worker.work(0)

    meta = MetaJob.get_meta(meta.meta_id)
    assert_equal MetaJob, meta.job_class
    assert meta.started?
    assert meta.finished?, 'Job should be finished'
    assert meta.succeeded?, 'Job should have succeeded'
    assert !meta.enqueued?
    assert meta.seconds_enqueued > 0.0, "seconds_enqueued should be greater than zero"
    assert meta.seconds_processing > 0.0, "seconds_processing should be greater than zero"
    assert_equal 'bar', meta['foo'], "'foo' not found in #{meta.inspect}"
  end

  def test_wrong_id_for_class
    meta = MetaJob.enqueue('foo', 'bar')

    assert_nil AnotherJob.get_meta(meta.meta_id)
    assert_not_nil Resque::Plugins::Meta.get_meta(meta.meta_id)
  end

  def test_expired_metadata
    meta = MetaJob.enqueue('foo', 'bar')
    worker = Resque::Worker.new(:test)
    worker.work(0)

    sleep 2
    meta = MetaJob.get_meta(meta.meta_id)
    assert_nil meta
  end

  def test_expired_metadata_before_finished
    meta = TransientJob.enqueue()
    worker = Resque::Worker.new(:test)
    assert meta = TransientJob.get_meta(meta.meta_id)

    sleep 2
    meta = TransientJob.get_meta(meta.meta_id)
    assert_equal nil, meta
  end

  def test_slow_job
    meta = SlowJob.enqueue('foo', 'bar')
    worker = Resque::Worker.new(:test)
    thread = Thread.new { worker.work(0) }

    sleep 0.1
    meta = SlowJob.get_meta(meta.meta_id)
    assert !meta.enqueued?
    assert meta.started?
    assert meta.working?
    assert !meta.finished?

    thread.join # job should be done
    meta.reload!
    assert !meta.enqueued?
    assert meta.started?
    assert !meta.working?
    assert meta.finished?
    assert meta.succeeded?
    assert !meta.failed?

    sleep 2
    assert_nil Resque::Plugins::Meta.get_meta(meta.meta_id)
  end

  def test_failing_job
    meta = FailingJob.enqueue()
    assert_nil meta.failed?
    worker = Resque::Worker.new(:test)
    worker.work(0)

    meta.reload!
    assert meta.finished?
    assert meta.failed?
    assert !meta.succeeded?
  end

  def test_saving_additional_metadata
    meta = MetaJob.enqueue('stuff')
    meta['foo'] = 'bar'
    meta.save

    # later
    meta = MetaJob.get_meta(meta.meta_id)
    assert_equal 'bar', meta['foo']
  end
end


test/test_helper.rb
dir = File.dirname(File.expand_path(__FILE__))
$LOAD_PATH.unshift dir + '/../lib'
require 'test/unit'
require 'rubygems'
require 'resque'

#
# make sure we can run redis
#

if !system("which redis-server")
  puts '', "** can't find `redis-server` in your path"
  puts "** try running `sudo rake install`"
  abort ''
end


#
# start our own redis when the tests start,
# kill it when they end
#

at_exit do
  next if $!

  exit_code = Test::Unit::AutoRunner.run

  pid = `ps -A -o pid,command | grep [r]edis-test`.split(" ")[0]
  puts "Killing test redis server[#{pid}]..."
  `rm -f #{dir}/dump.rdb`
  Process.kill("KILL", pid.to_i)
  exit exit_code
end

puts "Starting redis for testing at localhost:9736..."
`redis-server #{dir}/redis-test.conf`
Resque.redis = 'localhost:9736'


lib/resque/plugins/meta/metadata.rb
require 'time'

module Resque
  module Plugins
    module Meta
      class Metadata
        include Resque::Helpers
        attr_reader :job_class, :meta_id, :data, :enqueued_at, :expire_in, :before_finish_expire_in

        def initialize(data_hash)
          data_hash['enqueued_at'] ||= to_time_format_str(Time.now)
          @data = data_hash
          @meta_id = data_hash['meta_id'].dup
          @enqueued_at = from_time_format_str('enqueued_at')
          @job_class = data_hash['job_class']
          if @job_class.is_a?(String)
            @job_class = constantize(data_hash['job_class'])
          else
            data_hash['job_class'] = @job_class.to_s
          end
          @expire_in = @job_class.expire_meta_in || 0
          @before_finish_expire_in = @job_class.before_finish_expire_in || 0
        end

        # Reload the metadata from the store
        def reload!
          if new_meta = job_class.get_meta(meta_id)
            @data = new_meta.data
          end
          self
        end

        # Save the metadata
        def save
          job_class.store_meta(self)
          self
        end

        def [](key)
          data[key]
        end

        def []=(key, val)
          data[key.to_s] = val
        end

        def start!
          @started_at = Time.now
          self['started_at'] = to_time_format_str(@started_at)
          save
        end

        def started_at
          @started_at ||= from_time_format_str('started_at')
        end

        def finish!
          data['succeeded'] = true unless data.has_key?('succeeded')
          @finished_at = Time.now
          self['finished_at'] = to_time_format_str(@finished_at)
          save
        end

        def finished_at
          @finished_at ||= from_time_format_str('finished_at')
        end

        def expire_at
          # expiry after finished
          if finished? && expire_in > 0
            finished_at.to_i + expire_in
          # expiry after enqueued
          elsif before_finish_expire_in > 0
            enqueued_at.to_i + before_finish_expire_in
          # default ttl is forever
          else
            0
          end
        end

        def enqueued?
          !started?
        end

        def working?
          started? && !finished?
        end

        def started?
          started_at ? true :false
        end

        def finished?
          finished_at ? true : false
        end

        def fail!
          self['succeeded'] = false
          finish!
        end

        def succeeded?
          finished? ? self['succeeded'] : nil
        end

        def failed?
          finished? ? !self['succeeded'] : nil
        end

        def seconds_enqueued
          (started_at || Time.now).to_f - enqueued_at.to_f
        end

        def seconds_processing
          if started?
            (finished_at || Time.now).to_f - started_at.to_f
          else
            0
          end
        end

        protected

        def from_time_format_str(key)
          (t = self[key]) && Time.parse(t)
        end

        def to_time_format_str(time)
          time.utc.iso8601(6)
        end
      end
    end
  end
end


lib/resque/plugins/meta.rb
require 'digest/sha1'
require 'resque'
require 'resque/plugins/meta/version'
require 'resque/plugins/meta/metadata'

module Resque
  module Plugins
    # If you want to be able to add metadata for a job
    # to track anything you want, extend it with this module.
    #
    # For example:
    #
    #     require 'resque-meta'
    #
    #     class MyJob
    #       extend Resque::Plugins::Meta
    #
    #       def self.perform(meta_id, *args)
    #         heavy_lifting
    #       end
    #     end
    #
    #     meta0 = MyJob.enqueue('stuff')
    #     meta0.enqueued_at # => 'Wed May 19 13:42:41 -0600 2010'
    #     meta0.meta_id # => '03c9e1a045ad012dd20500264a19273c'
    #     meta0['foo'] = 'bar' # => 'bar'
    #     meta0.save
    #
    #     # later
    #     meta1 = MyJob.get_meta('03c9e1a045ad012dd20500264a19273c')
    #     meta1.job_class # => MyJob
    #     meta1.enqueued_at # => 'Wed May 19 13:42:41 -0600 2010'
    #     meta1['foo'] # => 'bar'
    module Meta

      # Override in your job to control the metadata id. It is
      # passed the same arguments as `perform`, that is, your job's
      # payload.
      def meta_id(*args)
        Digest::SHA1.hexdigest([ Time.now.to_f, rand, self, args ].join)
      end

      # Override in your job to control the how many seconds a job's
      # metadata will live after it finishes.  Defaults to 24 hours.
      # Return nil or 0 to set them to never expire.
      def expire_meta_in
        24 * 60 * 60
      end

      # Override in your job to control how many seconds a job's
      # metadata will live after it is enqueued and *before* it finishes.
      # This includes the condition
      # where the job and metadata are never finished because of a
      # shutdown or server crash. Defaults to 0 (i.e.
      # forever).  Return nil or 0 to set them to never expire.
      # Use caution: this value should be longer than your jobs take to run!
      def before_finish_expire_in
        0
      end

      # Enqueues a job in Resque and return the association metadata.
      # The meta_id in the returned object can be used to fetch the
      # metadata again in the future.
      def enqueue(*args)
        meta = Metadata.new({'meta_id' => meta_id(args), 'job_class' => self.to_s})
        meta.save
        Resque.enqueue(self, meta.meta_id, *args)
        meta
      end

      def store_meta(meta)
        key = "meta:#{meta.meta_id}"
        json = Resque.encode(meta.data)
        if meta.expire_at > 0
          Resque.redis.setex(key, meta.expire_at - Time.now.to_i, json)
        else
          Resque.redis.set(key, json)
        end
        meta
      end

      # Retrieve the metadata for a given job.  If you call this
      # from a class that extends Meta, then the metadata will
      # only be returned if the metadata for that id is for the
      # same class.  Explicitly, calling Meta.get_meta(some_id)
      # will return the metadata for a job of any type.
      def get_meta(meta_id)
        key = "meta:#{meta_id}"
        if json = Resque.redis.get(key)
          data = Resque.decode(json)
          if self == Meta || self.to_s == data['job_class']
            Metadata.new(data)
          end
        end
      end
      module_function :get_meta
      public :get_meta

      def before_perform_meta(meta_id, *args)
        if meta = get_meta(meta_id)
          meta.start!
        end
      end

      def after_perform_meta(meta_id, *args)
        if meta = get_meta(meta_id)
          meta.finish!
        end
      end

      def on_failure_meta(e, meta_id, *args)
        if meta = get_meta(meta_id)
          meta.fail!
        end
      end
    end
  end
end
